### PR TITLE
Bump bliss

### PR DESCRIPTION
**Pull request by matbesancon**
Bliss version 0.77 was released, with notably a cmake system which should drastically simplify the build recipe

### CODE REVIEW

**Comment by matbesancon** [comment]
Current error I'm getting on Mac ARM:
```
/usr/bin/cmake -E cmake_link_script CMakeFiles/bliss.dir/link.txt --verbose=false
Undefined symbols for architecture arm64:
  "___gmpz_clear", referenced from:
      _bliss_find_automorphisms in bliss_C.cc.o
      _bliss_find_canonical_labeling in bliss_C.cc.o
  "___gmpz_init", referenced from:
      _bliss_find_automorphisms in bliss_C.cc.o
      _bliss_find_canonical_labeling in bliss_C.cc.o
  "___gmpz_mul_ui", referenced from:
      __ZN5bliss13AbstractGraph6searchEbRNS_5StatsERKNSt3__18functionIFvjPKjEEERKNS4_IFbvEEE in abstractgraph.cc.o
  "___gmpz_set_ui", referenced from:
      __ZN5bliss13AbstractGraph6searchEbRNS_5StatsERKNSt3__18functionIFvjPKjEEERKNS4_IFbvEEE in abstractgraph.cc.o
      _bliss_find_automorphisms in bliss_C.cc.o
      _bliss_find_canonical_labeling in bliss_C.cc.o
ld: symbol(s) not found for architecture arm64
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
```
linked to GMP?

File: B/bliss/build_tarballs.jl [hunk 1]
```diff
@@ -3,30 +3,29 @@
...
 
 # Bash recipe for building across all platforms
 script = raw"""
 cd $WORKSPACE/srcdir
-cd bliss-0.73
-if [[ "${target}" == *mingw* ]]; then
-  atomic_patch -p1 ../patches/notimer.patch
-fi
-# build with GMP and store this information
-sed -i -e 's/^namespace bliss/#define BLISS_USE_GMP\n\nnamespace bliss/' defs.hh
```
Comment (benlorenz) [inline B/bliss/build_tarballs.jl:22]: Don't forget to install the headers as well and please keep this line or at least something adapted from it. Otherwise it will be very difficult for any downstream packages to know whether bliss was built with or without GMP (and the wrong choice will produce segfaults at runtime), this just inserts that definition into one main header file and will make sure anything depending on bliss also uses that flag during compilation.

Comment (matbesancon) [reply B/bliss/build_tarballs.jl:22]: thanks for the headers. For GMP, it seems the new version already adds this flag to defs.hh
```
/workspace/srcdir/bliss-0.77/src/defs.hh:33:9: warning: 'BLISS_USE_GMP' macro redefined [-Wmacro-redefined]
#define BLISS_USE_GMP
        ^
<command line>:3:9: note: previous definition is here
#define BLISS_USE_GMP 1
        ^
1 warning generated.
```

File: B/bliss/build_tarballs.jl [hunk 2]
```diff
@@ -3,35 +3,37 @@
...
+  -DUSE_GMP=ON ..
+make -j${nproc}
+cp bliss$exeext $bindir
+cp libbliss.${dlext} $libdir
+
 mkdir -p $prefix/include/bliss
-mkdir -p $libdir
-install -p -m 0644 -t $prefix/include/bliss *.hh
-install -p libbliss.$dlext $libdir
+install -p -m 0644 -t $prefix/include/bliss $WORKSPACE/srcdir/bliss-*/src/*.hh
```
Comment (giordano) [inline B/bliss/build_tarballs.jl:28]: It's always better to use the most specic environment variable
```suggestion
install -p -m 0644 -t "${includedir}/bliss" ../src/*.hh
```

File: B/bliss/build_tarballs.jl [hunk 3]
```diff
@@ -3,35 +3,37 @@
...
+
+mkdir -p ${prefix}/share/licenses/bliss
+cp $WORKSPACE/srcdir/bliss-*/COPYING.LESSER ${prefix}/share/licenses/bliss/LICENSE_bliss
```
Comment (giordano) [inline B/bliss/build_tarballs.jl:31]: Any specific reason for not doing
```suggestion
install_license ../COPYING.LESSER
```
?

**Comment by matbesancon** [comment]
For the moment the error still occurring (on ARM at least):
```
/opt/bin/aarch64-apple-darwin20-libgfortran5-cxx11/aarch64-apple-darwin20-clang++ --sysroot=/opt/aarch64-apple-darwin20/aarch64-apple-darwin20/sys-root    -O3 -DNDEBUG   -Wall -pedantic -O3 -DNDEBUG -DBLISS_USE_GMP -I/opt/aarch64-apple-darwin20/aarch64-apple-darwin20/sys-root/usr/local/include -std=gnu++11 -o CMakeFiles/bliss-executable.dir/src/bliss.cc.o -c /workspace/srcdir/bliss-0.77/src/bliss.cc
Undefined symbols for architecture arm64:
  "___gmpz_clear", referenced from:
      _bliss_find_automorphisms in bliss_C.cc.o
      _bliss_find_canonical_labeling in bliss_C.cc.o
  "___gmpz_init", referenced from:
      _bliss_find_automorphisms in bliss_C.cc.o
      _bliss_find_canonical_labeling in bliss_C.cc.o
  "___gmpz_mul_ui", referenced from:
      __ZN5bliss13AbstractGraph6searchEbRNS_5StatsERKNSt3__18functionIFvjPKjEEERKNS4_IFbvEEE in abstractgraph.cc.o
  "___gmpz_set_ui", referenced from:
      __ZN5bliss13AbstractGraph6searchEbRNS_5StatsERKNSt3__18functionIFvjPKjEEERKNS4_IFbvEEE in abstractgraph.cc.o
      _bliss_find_automorphisms in bliss_C.cc.o
      _bliss_find_canonical_labeling in bliss_C.cc.o
ld: symbol(s) not found for architecture arm64
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [CMakeFiles/bliss.dir/build.make:224: libbliss.dylib] Error 1
make[2]: Leaving directory '/workspace/srcdir/bliss-0.77/build'
make[1]: *** [CMakeFiles/Makefile2:154: CMakeFiles/bliss.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[100%] Linking CXX executable bliss
```

File: B/bliss/build_tarballs.jl (hunk 1 above)
Comment (benlorenz) [reply B/bliss/build_tarballs.jl:22]:
```
/workspace/srcdir/bliss-0.77/src/defs.hh:33:9: warning: 'BLISS_USE_GMP' macro redefined [-Wmacro-redefined]
#define BLISS_USE_GMP
        ^
<command line>:3:9: note: previous definition is here
#define BLISS_USE_GMP 1
        ^
1 warning generated.
```
This warning is a conflict between a flag from the command line and the one added to defs.hh. Setting `-DUSE_GMP=on` for cmake seems to add `-DBLISS_USE_GMP=1` to the CFLAGS, but this is not stored anywhere for later usage. To avoid the warning you could patch this after building the library.

Comment (matbesancon) [reply B/bliss/build_tarballs.jl:22]: couldn't we just use `#ifdef BLISS_USE_GMP`?

**Comment by benlorenz** [comment]
Please try something like this for gmp:
```diff
--- bliss-0.77/CMakeLists.txt.orig	2021-12-06 20:41:29.264752974 +0100
+++ bliss-0.77/CMakeLists.txt	2021-12-06 20:41:48.839856607 +0100
@@ -59,6 +59,7 @@
 add_executable(bliss-executable src/bliss.cc)
 target_link_libraries(bliss-executable bliss_static)
 if(USE_GMP)
+  target_link_libraries(bliss ${GMP_LIBRARIES})
   target_link_libraries(bliss-executable ${GMP_LIBRARIES})
 endif(USE_GMP)
 set_target_properties(bliss-executable PROPERTIES OUTPUT_NAME bliss)
```

The cmake file doesn't add gmp to the linker flags for the library, only for the binary. On Linux this doesn't give an error as the linker ignores undefined symbols by default and assumes that they can be resolved correctly at runtime.

File: B/bliss/build_tarballs.jl (hunk 1 above)
Comment (benlorenz) [reply B/bliss/build_tarballs.jl:22]: Something like the following would also work yes (and even avoid warnings if anyone downstream manually sets BLISS_USE_GMP).
```c
#ifndef BLISS_USE_GMP
#define BLISS_USE_GMP 1
#endif
```

(Most other software packages generate some config.h from a config.h.in which contains such options, or generate a pkg-config files which contains the flag in the CFLAGS)

**Comment by matbesancon** [comment]
seems all good now!

File: B/bliss/build_tarballs.jl [hunk 4]
```diff
@@ -3,35 +3,41 @@
...
-install -p libbliss.$dlext $libdir
+install -p -m 0644 -t "${includedir}/bliss" ../src/*.hh
+
+install_license ../COPYING.LESSER
 """
 
 # These are the platforms we will build for by default, unless further
 # platforms are passed in on the command line
-platforms = supported_platforms(;experimental=true)
+platforms = expand_cxxstring_abis(supported_platforms(;experimental=true))
```
Comment (giordano) [inline B/bliss/build_tarballs.jl:40]: I've just noticed: why this change?  I don't see any indication this was necessary.  I blocked registration because of this: https://github.com/JuliaRegistries/General/pull/51827

Comment (matbesancon) [reply B/bliss/build_tarballs.jl:40]: I think I got a warning at some point, but might be wrong, do you want a new PR removing it?

Comment (giordano) [reply B/bliss/build_tarballs.jl:40]: I opened #4195: I don't see the warning anywhere https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=15855&view=results

Comment (matbesancon) [reply B/bliss/build_tarballs.jl:40]: then my bad!
